lib/Dialect/Arc/Transforms/LatencyRetiming.cpp

```diff
@@ -0,0 +1,136 @@
+//===- LatencyRetiming.cpp - Implement LatencyRetiming Pass ---------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Support/LLVM.h"
+#include "circt/Support/SymCache.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "arc-latency-retiming"
+
+using namespace circt;
+using namespace arc;
+
+//===----------------------------------------------------------------------===//
+// Patterns
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct LatencyRetimingStatistics {
+  unsigned numOpsRemoved = 0;
+  unsigned latencyUnitsSaved = 0;
+};
+
+/// Absorb the latencies from predecessor states to collapse shift registers and
+/// reduce the overall amount of latency units in the design.
+struct LatencyRetimingPattern : OpRewritePattern<StateOp> {
+  LatencyRetimingPattern(MLIRContext *context, SymbolCache &symCache,
+                         LatencyRetimingStatistics &statistics)
+      : OpRewritePattern<StateOp>(context), symCache(symCache),
+        statistics(statistics) {}
+
+  LogicalResult matchAndRewrite(StateOp op,
+                                PatternRewriter &rewriter) const final;
+
+private:
+  SymbolCache &symCache;
+  LatencyRetimingStatistics &statistics;
+};
+
+} // namespace
+
+LogicalResult
+LatencyRetimingPattern::matchAndRewrite(StateOp op,
+                                        PatternRewriter &rewriter) const {
+  unsigned minPrevLatency = UINT_MAX;
+  SetVector<StateOp> predecessors;
+
+  for (auto input : op.getInputs()) {
+    auto predState = input.getDefiningOp<StateOp>();
+    if (!predState)
+      return failure();
+
+    if (predState->hasAttr("name") || predState->hasAttr("names"))
+      return failure();
+
+    if (predState == op)
+      return failure();
+
+    if (predState.getLatency() != 0 && op.getLatency() != 0 &&
+        predState.getClock() != op.getClock())
+      return failure();
+
+    if (predState.getEnable() || predState.getReset() || op.getReset() ||
+        op.getEnable())
+      return failure();
```

Comment:
Good call just bailing if any of these signals are present. My first reflex was "damn enable making everything tricky", but I think even if the enable were represented as a mux you still couldn't merge states because that would introduce a lot of self-loops. Enables seem to be inherently challenging, and better to have them represented without the mux and self-loop.

Nit: You could pull the check for `op.get{Reset,Enable}()` out to before the for loop for an early exit.

Reply:
Yeah, enables are quite annoying in that they hinder a lot of optimizations.

Excellent point! Pulled the check out of the loop.